src/showingpreviously/cinemas/picturehouse.py

```diff
@@ -0,0 +1,174 @@
+import json
+import re
+
+from bs4 import BeautifulSoup
+from datetime import datetime, timedelta
+
+import showingpreviously.requests as requests
+from showingpreviously.model import ChainArchiver, CinemaArchiverException, Chain, Cinema, Screen, Film, Showing
+from showingpreviously.consts import STANDARD_DAYS_AHEAD, UK_TIMEZONE
+
+
+PICTUREHOUSE_TOKEN_URL = 'https://www.picturehouses.com/'
+CINEMAS_API_URL = 'https://www.picturehouses.com/ajax-cinema-list'
+SHOWINGS_API_URL = 'https://www.picturehouses.com/api/scheduled-movies-ajax'
+FILM_INDEX_URL = 'https://www.picturehouses.com/whats-on'
+FILM_DETAILS_URL = 'https://www.picturehouses.com/movie-details/000/{id}/{slug}'
+
+CHAIN = Chain('Picturehouse')
+
+LARAVEL_PATTERN = re.compile(r'laravel_session=(?P<laravel_session>.+?);')
+SHOWING_URL_ID_PATTERN = re.compile(r'/movie-details/\d+/(?P<film_id>.+?)/(?P<slug>.+)')
+SLUG_YEAR_PATTERN = re.compile(r'(?P<year>(?:18|19|20)\d{2})')
+JS_TO_URL_PATTERN = re.compile(r'"(?P<film_link>https?://.+?)"')
+
+
+def get_showing_dates() -> str:
+    current_date = datetime.now()
+    end_date = current_date + timedelta(days=STANDARD_DAYS_AHEAD)
+    while current_date < end_date:
+        yield current_date.strftime('%Y-%m-%d')
+        current_date += timedelta(days=1)
+
+
+def get_tokens() -> (str, str):
+    r = requests.get(PICTUREHOUSE_TOKEN_URL)
+    if r.status_code != 200:
+        raise CinemaArchiverException(f'Got status code {r.status_code} when fetching URL {PICTUREHOUSE_TOKEN_URL}')
+    all_cookies = r.headers.get('set-cookie')
+    laravel_session = LARAVEL_PATTERN.search(all_cookies).group('laravel_session')
+    soup = BeautifulSoup(r.text, features='html.parser')
+    token = soup.find('input', {'name': '_token'})['value']
+    return laravel_session, token
+
+
+def get_cinemas(laravel_session: str, token: str) -> dict[str, Cinema]:
+    r = requests.post(CINEMAS_API_URL, cookies={'laravel_session': laravel_session}, data={'_token': token})
+    if r.status_code != 200:
+        raise CinemaArchiverException(f'Got status code {r.status_code} when fetching URL {CINEMAS_API_URL}')
+    try:
+        cinema_data = r.json()
+    except json.JSONDecodeError:
+        raise CinemaArchiverException(f'Error decoding JSON data from URL {CINEMAS_API_URL}')
+    cinemas = {}
+    for cinema in cinema_data['cinema_list']:
+        id = cinema['cinema_id']
+        name = cinema['name']
+        cinemas[id] = Cinema(name, UK_TIMEZONE)
+    return cinemas
+
+
+def get_film_year(film_link: str) -> str:
+    # first we inspect the url to see if the date is present in that
+    match = SLUG_YEAR_PATTERN.search(film_link)
+    if match:
+        return match.group('year')
+
+    # then we request the page and look for the date there
+    r = requests.get(film_link)
+    if r.status_code != 200:
+        # sometimes films don't exist for some reason. nothing we can do about that
+        return '0'
```

Comment:
We now have a constant that indicates when a year is not present. That should be used instead.